README.md
# ZenPacks.itri.NetworkTopology

Maps network topology in Zenoss using LLDP and D3.

## Usage

Switch devices are added to the `/Network/Switch` device class. Once they are added they will be modeled by Zenoss and the available network links will be discovered and modeled:

![Network Links](screenshots/network_links.jpg)

The network topology map can then be visualized by navigating to *Infrastructure > Network Map* and selecting a switch device from the menu and clicking *Apply*:

![Network Topology](screenshots/network_topology.jpg)

This ZenPack replaces Zenoss's original default *Network Map* page.

CHANGELOG.md
## Change Log

### 1.0.0
- Initial release

ZenPacks/itri/NetworkTopology/lib/lldp.py
import Globals
import sys, os
import logging
log = logging.getLogger('zen.NetworkTopology.lldp')

from pynetsnmp.SnmpSession import *
from pynetsnmp.netsnmp import *

# Error codes
ERR_DEVICE_TYPE_ERROR = 51101
ERR_SNMP_CONNECTION_FAILED = 51102
ERR_PARTIAL_SNMP_CONNECTION_FAILED = 51105

# OIDs
MIB_LLDP_REM_MAN_ADDR_IF_SUBTYPE = '.1.0.8802.1.1.2.1.4.2.1.3'
MIB_LLDP_REM_PORT_DESC = '.1.0.8802.1.1.2.1.4.1.1.8'
MIB_LLDP_REM_SYS_DESC = '.1.0.8802.1.1.2.1.4.1.1.10'
MIB_RFC1213_SYS_DESC = '.1.3.6.1.2.1.1.1'


class Link():
    """Class to represent links between a local and remote devices."""

    def __init__(self, local_type='Unknown', local_ip='Unknown', 
                 local_port='Unknown', remote_type='Unknown', 
                 remote_ip='Unknown', remote_port='Unknown'):
        self.local_type = local_type
        self.local_ip = local_ip
        self.local_port = local_port
        self.remote_type = remote_type
        self.remote_ip = remote_ip
        self.remote_port = remote_port

    def __repr__(self):
        return '{0}\t{1}\t{2}\t{3}\t{4}\t{5}'.format(self.local_type, self.local_ip, 
            self.local_port, self.remote_type, self.remote_ip, self.remote_port)

class LLDPSession(SnmpSession):
    """A SNMP Session to gather switch network topology information."""

    def get_remote_ip_and_local_port(self):
        """ Returns a list of tuples. Each tuple consists of a local port and a remote mgmIP."""
        ret = []

        remManAddrTable = self._walk(MIB_LLDP_REM_MAN_ADDR_IF_SUBTYPE)
        
        for address in remManAddrTable:
            ip = ''
            for key in address.keys()[0][16:20]:
                ip = ip + '.' + str(key)
            ret.append((address.keys()[0][12], ip[1:]))
        
        return ret

    def get_remote_port_description(self):
        """ return a dictionary with an index of local port and a value of remote port """
        ret = {}

        remote_ports_table = self._walk(MIB_LLDP_REM_PORT_DESC)

        for i in remote_ports_table:
            port = i.values()[0].split(' ')
            if len(port) == 1:   # server
                ret[i.keys()[0][12]] = i.values()[0]
            else:                # switch
                for x in range(0, len(port)-1):
                    if port[x] == 'port' or port[x] == 'Port':
                        ret[i.keys()[0][12]] = 'port' + port[x+1]
        return ret

    def get_remote_device_type(self):
        """ return a dictionary with an index of local port and a value of remote device type """
        ret = {}

        remDeviceTable = self._walk(MIB_LLDP_REM_SYS_DESC)
        for device in remDeviceTable:
            type = device.values()[0].split(' ')
            if type[0] == 'Linux':
                ret[device.keys()[0][-2]] = 'Server'
            else:
                ret[device.keys()[0][-2]] = 'Switch'

        return ret

    def get_local_device_type(self):
        """ Returns the device type of this session."""
        snmp_result = self.get(MIB_RFC1213_SYS_DESC + '.0')
        device_type = snmp_result.values()[0].split(' ')

        if device_type[0] == 'Linux':
            return 'Server'

        return 'Switch'

    def _walk(self, root_oid):
        table_result = []
        
        current_tuple = self._getNext(root_oid)
        base_oid = tuple(map(int, (root_oid).strip('.').split('.')))
        
        while True:
            if current_tuple.keys()[0][:len(base_oid)] == base_oid:
                table_result.append(current_tuple)
                oid = ''
                
                for key in current_tuple.keys()[0]:
                    oid = oid + '.' + str(key)
                current_tuple = self._getNext(oid)
            else:
                break
                
        return table_result

    def _getNext(self, oid):
        self.session = netsnmp.Session(
            version = self._version,
            timeout = int(self.timeout*1e6),
             retries=int(self.retries-1),
            peername= '%s:%d' % (self.ip, self.port),
            community=self.community,
            community_len=len(self.community),
            cmdLineArgs=self.cmdLineArgs
            )

        oid = tuple(map(int, oid.strip('.').split('.')))
        self.session.open()
        try:
            return self._sgetnext([oid])
        finally:
            self.session.close()


    def _sgetnext(self, oids):
        req = self.session._create_request(SNMP_MSG_GETNEXT)

        for oid in oids:
            oid = mkoid(oid)
            lib.snmp_add_null_var(req, oid, len(oid))

        response = netsnmp_pdu_p()

        if lib.snmp_synch_response(self.session.sess, req, byref(response)) == 0:
            result = dict(getResult(response.contents))
            lib.snmp_free_pdu(response)
            return result

def get_switch_links(switch_ip):
    session = LLDPSession(switch_ip, timeout=1.5, port=161)
    session.community = 'public'
    
    try:
        if session.get_local_device_type() == 'Server':
            return ERR_DEVICE_TYPE_ERROR, [], [switch_ip]
    except AttributeError:
        log.error("Error: SNMP connection failed.")
        return ERR_SNMP_CONNECTION_FAILED, [], [switch_ip]
    except Exception as e:
        log.error("Unknown Error: {0}".format(e.message))
        return ERR_UNKNOWN_ERROR, [], [switch_ip]
        
    todo_switches = [switch_ip]
    processed_switches = []
    failed_switches = []

    links = []    
    for switch in todo_switches:
        if switch not in processed_switches:
            session = LLDPSession(switch, timeout=1.5, port=161)
            session.community = 'public'
        
            try:
                processed_switches.append(switch)
                
                remote_ips_table = session.get_remote_ip_and_local_port()                
                remote_ports_table = session.get_remote_port_description()
                remote_device_types = session.get_remote_device_type()
                
                for remote_ip in remote_ips_table:
                    link = Link(local_type='Switch', local_ip=switch, local_port=('port' + str(remote_ip[0])))
                    
                    # Assign the remote device's type to the link
                    if remote_device_types[remote_ip[0]]:
                        link.remote_type = remote_device_types[remote_ip[0]]
                        
                    link.remote_ip = remote_ip[1]
                    
                    if remote_ports_table[remote_ip[0]]:
                        link.remote_port = remote_ports_table[remote_ip[0]]
                        
                    links.append(link)
                    
                    if link.remote_type == 'Switch':
                        todo_switches.append(link.remote_ip)
            except Exception as e:
                log.error("Error obtaining switch links for {0}: {1}".format(switch_ip, e.message))
                failed_switches.append(switch)

    return links

if __name__ == '__main__':
    links = get_switch_links("100.67.0.21")
    for link in links:
        print link


ZenPacks/itri/NetworkTopology/browser/resources/js/graph.js
(function() {
  function render_graph(panel_selector, on_node_click) {

    var links = [
      {source: "Microsoft", target: "Amazon", type: "licensing"},
      {source: "Microsoft", target: "HTC", type: "licensing"},
      {source: "Samsung", target: "Apple", type: "suit"},
      {source: "Motorola", target: "Apple", type: "suit"},
      {source: "Nokia", target: "Apple", type: "resolved"},
      {source: "HTC", target: "Apple", type: "suit"},
      {source: "Kodak", target: "Apple", type: "suit"},
      {source: "Microsoft", target: "Barnes & Noble", type: "suit"},
      {source: "Microsoft", target: "Foxconn", type: "suit"},
      {source: "Oracle", target: "Google", type: "suit"},
      {source: "Apple", target: "HTC", type: "suit"},
      {source: "Microsoft", target: "Inventec", type: "suit"},
      {source: "Samsung", target: "Kodak", type: "resolved"},
      {source: "LG", target: "Kodak", type: "resolved"},
      {source: "RIM", target: "Kodak", type: "suit"},
      {source: "Sony", target: "LG", type: "suit"},
      {source: "Kodak", target: "LG", type: "resolved"},
      {source: "Apple", target: "Nokia", type: "resolved"},
      {source: "Qualcomm", target: "Nokia", type: "resolved"},
      {source: "Apple", target: "Motorola", type: "suit"},
      {source: "Microsoft", target: "Motorola", type: "suit"},
      {source: "Motorola", target: "Microsoft", type: "suit"},
      {source: "Huawei", target: "ZTE", type: "suit"},
      {source: "Ericsson", target: "ZTE", type: "suit"},
      {source: "Kodak", target: "Samsung", type: "resolved"},
      {source: "Apple", target: "Samsung", type: "suit"},
      {source: "Kodak", target: "RIM", type: "suit"},
      {source: "Nokia", target: "Qualcomm", type: "suit"}
    ];

    var nodes = {};

    // Compute the distinct nodes from the links.
    links.forEach(function(link) {
      link.source = nodes[link.source] || (nodes[link.source] = {name: link.source});
      link.target = nodes[link.target] || (nodes[link.target] = {name: link.target});
    });

    var width = 960,
        height = 500;

    var force = d3.layout.force()
        .nodes(d3.values(nodes))
        .links(links)
        .size([width, height])
        .linkDistance(60)
        .charge(-300)
        .on("tick", tick)
        .start();

    var svg = d3.select("#" + panel_selector.body.id).append("svg")
        .attr("width", width)
        .attr("height", height);

    // Per-type markers, as they don't inherit styles.
    svg.append("defs").selectAll("marker")
        .data(["suit", "licensing", "resolved"])
      .enter().append("marker")
        .attr("id", function(d) { return d; })
        .attr("viewBox", "0 -5 10 10")
        .attr("refX", 15)
        .attr("refY", -1.5)
        .attr("markerWidth", 6)
        .attr("markerHeight", 6)
        .attr("orient", "auto")
      .append("path")
        .attr("d", "M0,-5L10,0L0,5");

    var path = svg.append("g").selectAll("path")
        .data(force.links())
      .enter().append("path")
        .attr("class", function(d) { return "link " + d.type; })
        .attr("marker-end", function(d) { return "url(#" + d.type + ")"; });

    var circle = svg.append("g").selectAll("circle")
        .data(force.nodes())
      .enter().append("circle")
        .attr("r", 6)
        .call(force.drag);

    var text = svg.append("g").selectAll("text")
        .data(force.nodes())
      .enter().append("text")
        .attr("x", 8)
        .attr("y", ".31em")
        .text(function(d) { return d.name; });

    // Use elliptical arc path segments to doubly-encode directionality.
    function tick() {
      path.attr("d", linkArc);
      circle.attr("transform", transform);
      text.attr("transform", transform);
    }

    function linkArc(d) {
      var dx = d.target.x - d.source.x,
          dy = d.target.y - d.source.y,
          dr = Math.sqrt(dx * dx + dy * dy);
      return "M" + d.source.x + "," + d.source.y + "A" + dr + "," + dr + " 0 0,1 " + d.target.x + "," + d.target.y;
    }

    function transform(d) {
      return "translate(" + d.x + "," + d.y + ")";
    }

  };

})();


ZenPacks/itri/NetworkTopology/modeler/plugins/itri/NetworkLinks.py
"""Models the network links for a switch device using LLDP."""

from twisted.internet.defer import inlineCallbacks, returnValue

from Products.DataCollector.plugins.CollectorPlugin import PythonPlugin

from ZenPacks.itri.NetworkTopology.lib import lldp


class NetworkLinks(PythonPlugin):
    relname = 'networkLinks'
    modname = 'ZenPacks.itri.NetworkTopology.NetworkLink'

    requiredProperties = ()

    deviceProperties = PythonPlugin.deviceProperties + requiredProperties
    
    @inlineCallbacks
    def collect(self, device, log):
        log.info('Collecting network link data for switch device {0}'.format(
            device.id))

        try:
            switch_links = yield lldp.get_switch_links(device.id)
        except Exception as e:
            log.error('{0}: {1}'.format(device.id, e))

        rm = self.relMap()
        
        for link in switch_links:
            rm.append(self.objectMap({
                'id': self.prepId(link.local_ip + '_' + link.local_port),
                'local_ip': link.local_ip,
                'local_type': link.local_type,
                'local_port': link.local_port,
                'remote_ip':  link.remote_ip,
                'remote_type': link.remote_type,
                'remote_port': link.remote_port,
                }))
        
        returnValue(rm)
    
    def process(self, device, results, log):
        log.info('Processing network link data for switch device {0}'.format(
            device.id))
        log.debug(results)
        
        return results


ZenPacks/itri/NetworkTopology/services/NetworkTopologyConfigService.py
"""
NetworkTopologyConfigService
Zenhub service for providing configuration to the nmapper collector daemon.
"""

import logging
log = logging.getLogger('zen.nmapper')

import Globals
from Products.ZenUtils.Utils import unused
from Products.ZenCollector.services.config import CollectorConfigService

unused(Globals)

class NetworkTopologyConfigService(CollectorConfigService):
    """ Zenhub service for nmapper daemon."""
    
    def _filterDevice(self, device):
        filter = CollectorConfigService._filterDevice(self, device)
        
        has_flag = False
        if filter:
            try:
                if '/Network/Switch' in device.getDeviceClassName():
                    has_flag = True
            except Exception as e:
                log.error(e.message)
        
        return CollectorConfigService._filterDevice(self, device) and has_flag
        
    def _createDeviceProxy(self, device):
        proxy = CollectorConfigService._createDeviceProxy(self, device)
        
        proxy.configCycleInterval = 5 * 60
        proxy.datapoints = []
        
        perfServer = device.getPerformanceServer()
        
        return proxy
        
if __name__ == '__main__':
    from Products.ZenHub.ServiceTester import ServiceTester
    
    tester = ServiceTester(NetworkTopologyConfigService)
    
    def printer(config):
        print config.datapoints
    
    tester.printDeviceProxy = printer
    tester.showDeviceInfo()


ZenPacks/itri/NetworkTopology/browser/resources/js/render_topology_map.js
Ext.onReady(function(){

  // Error Flares
  var show_error = Zenoss.flares.Manager.error;

  var map_panel = Ext.create('Ext.panel.Panel', {
    flex: 1
  });

  var refresh_map = function() {
    d3.select("svg").remove();
    render_map(map_panel, sidebar.getValues()['root_id']);
  };


  var sidebar = Ext.create('Ext.form.Panel', {
    id: 'network_topology_form',
    width: 300,
    bodyPadding: 10,
    frame: true,
    defaultType: 'textfield',
    layout: {
      type: 'vbox',
      padding: 5,
      align: 'stretch'
    },
    items: [
      {
        id: 'sidebar_switch_id_label',
        name: 'switch_id_label',
        value: 'Switch IP',
        xtype: 'displayfield'
      },
      {
        id: 'sidebar_switch_id',
        fieldlabel: 'Switch ID',
        name: 'root_id',
        xtype: 'combo',
        valueField: 'name',
        displayField: 'name',
        store: new Ext.data.DirectStore({
          directFn: Zenoss.remote.DeviceRouter.getDeviceUuidsByName,
          root: 'data',
          model: 'Zenoss.model.BasicUUID',
          remoteFilter: true
        }),
        minChars: 3,
        typeAhead: false,
        hideLabel: true,
        hideTrigger: false,
        listConfig: {
          loadingText: 'Searching...',
          emptyText: 'No matching switches found.'
        },
        pageSize: 10
      },
      {
        text: 'Apply',
        name: 'refresh_button',
        xtype: 'button',
        handler: refresh_map
      }
    ]
  });

  var hbox_center_panel = Ext.create('Ext.panel.Panel', {
    layout: {
      type: 'hbox',
      pack: 'start',
      align: 'stretch'
    }
  });
 
  hbox_center_panel.add(sidebar);
  hbox_center_panel.add(map_panel);    
  Ext.getCmp('center_panel').add(hbox_center_panel);

  function render_map(panel, switch_ip) {

  // First get the switch data
  Zenoss.remote.DeviceRouter.getComponents({uid: '/zport/dmd/Devices/Network/Switch/devices/' + switch_ip, limit: 1000, keys: ['local_ip', 'remote_ip', 'remote_type']}, function(result) {
    var width = 1600,
        height = 900;

    var force = d3.layout.force()        
        .size([width, height])
        .linkDistance(200)
        .charge(-800)
        .gravity(0.06)

    var drag = force.drag()
        .on("dragstart", dragstart);

    var svg = d3.select("#" + panel.body.id).append("svg")
        .attr("width", width)
        .attr("height", height);

    var links = result.data;
    var nodes = {};

    // Compute the distinct nodes from the links.
    links.forEach(function(link) {
      link.source = nodes[link.local_ip] || (nodes[link.local_ip] = {name: link.local_ip, type: "switch"});
      link.target = nodes[link.remote_ip] || (nodes[link.remote_ip] = {name: link.remote_ip, type: link.remote_type});
    });

    force
          .links(links)
          .nodes(d3.values(nodes))
          .on("tick", tick)
          .start();

    var link = svg.selectAll(".link")
        .data(force.links())
      .enter().append("line")
        .attr("class", "link");

    var node = svg.selectAll(".node")
        .data(force.nodes())
      .enter().append("g")
        .attr("class", "node")
        .on("dblclick", dblclick)
        .call(drag);

    node.append("image")
        .attr("x", -16)
        .attr("y", -16)
        .attr("width", 32)
        .attr("height", 32)
        .attr("xlink:href", function(d) { return "/++resource++networktopology/img/" + d.type.toLowerCase() + ".png"; });

    node.append("text")
        .attr("dx", -30)
        .attr("dy", 30)
        .text(function(d) { return d.name; });

    function tick() {
      link.attr("x1", function(d) { return d.source.x; })
          .attr("y1", function(d) { return d.source.y; })
          .attr("x2", function(d) { return d.target.x; })
          .attr("y2", function(d) { return d.target.y; });

      node.attr("transform", function(d) { return "translate(" + d.x + "," + d.y + ")"; });
    }
    
    function dragstart(d) {
      d3.select(this).classed("fixed", d.fixed = true);
    }
    
    function dblclick(d) {
      d3.select(this).classed("fixed", d.fixed = false);
    }
  });
  }
});
